File: PY/Bryann Valderrama/Strings/anagramsPattern.py
from sys import stdout
NO_OF_CHARS = 256


def compare(arr1, arr2):
    global NO_OF_CHARS
    for i in range(NO_OF_CHARS):
        if arr1[i] != arr2[i]:
            return False
    return True


def anagramsSearch(pat, txt):
    global NO_OF_CHARS
    M = len(pat)
    N = len(txt)
    countP = [0 for x in range(NO_OF_CHARS)]
    countTW = [0 for x in range(NO_OF_CHARS)]
    for i in range(M):
        countP[ord(pat[i])] += 1
        countTW[ord(txt[i])] += 1
    for i in range(M, N):
        if compare(countP, countTW):
            stdout.write(
                f'Encontrado en indice {i-M} | Anagrama: {txt[i-M:i]}\n')
        countTW[ord(txt[i])] += 1
        countTW[ord(txt[i-M])] -= 1
    if compare(countP, countTW):
        stdout.write(
            f'Encontrado en indice {N-M} | Anagrama: {txt[N-M:N]}\n')


# String
txt = 'lalan'

# *-------- Una Palabra --------*
stdout.write('Una Palabra\n')
pat = 'la'
anagramsSearch(pat, txt)

# *-------- Lista de Palabras --------*
stdout.write('Lista de Palabras\n')
pat = ['la', 'na', 'a']
for i in pat:
    anagramsSearch(i, txt)

File: C++/Bryann Valderrama/Strings/anagramsPattern.cpp
#include <cstdlib>
#include <bits/stdc++.h>
using namespace std;
#define NO_OF_CHARS 256

bool compare(char arr1[], char arr2[])
{
    for (int i = 0; i < NO_OF_CHARS; i++)
    {
        if (arr1[i] != arr2[i])
        {
            return false;
        }
    }
    return true;
}

void anagramsSearch(string pat, string txt)
{
    int M = pat.size();
    int N = txt.size();
    char countP[NO_OF_CHARS];
    char countTW[NO_OF_CHARS];
    //Inicializar los arreglos
    memset(countP, 0, sizeof countP);
    memset(countTW, 0, sizeof countTW);
    for (int i = 0; i < M; i++)
    {
        countP[pat[i]]++;
        countTW[txt[i]]++;
    }
    for (int i = M; i < N; i++)
    {
        if (compare(countP, countTW))
        {
            cout << "Encontrado en indice " << i - M << " | Anagrama: " << txt.substr(i - M, M) << endl;
        }
        countTW[txt[i]]++;
        countTW[txt[i - M]]--;
    }
    if (compare(countP, countTW))
    {
        cout << "Encontrado en indice " << N - M << " | Anagrama: " << txt.substr(N - M, N) << endl;
    }
}

int main(int argc, char const *argv[])
{
    string txt = "anitalavalatina";
    string pat = "la";
    anagramsSearch(pat, txt);
    return 0;
}
